linera-core/src/updater.rs

```diff
@@ -241,7 +242,47 @@ where
             .collect::<Result<Vec<_>, _>>()?;
             result = self
                 .node
-                .handle_certificate(certificate.clone(), values, delivery)
+                .handle_certificate(certificate.clone(), values.clone(), vec![], delivery)
+                .await;
+            values
+        } else {
+            vec![]
+        };
+
+        if let Err(NodeError::BlobsNotFound(blob_ids)) = &result {
+            // Find the missing blobs locally and retry.
+            let required = match certificate.value() {
+                CertificateValue::ConfirmedBlock { executed_block, .. }
+                | CertificateValue::ValidatedBlock { executed_block, .. } => {
+                    executed_block.block.blob_ids()
+                }
+                CertificateValue::Timeout { .. } => HashSet::new(),
+            };
+            for blob_id in blob_ids {
+                if !required.contains(blob_id) {
+                    warn!(
+                        "validator requested blob {:?} but it is not required",
+                        blob_id
+                    );
+                    return Err(NodeError::InvalidChainInfoResponse);
+                }
+            }
+            let unique_blob_ids = blob_ids.iter().cloned().collect::<HashSet<_>>();
+            if blob_ids.len() > unique_blob_ids.len() {
+                warn!("blobs requested by validator contain duplicates");
+                return Err(NodeError::InvalidChainInfoResponse);
+            }
+            let blobs = future::join_all(
+                unique_blob_ids
+                    .into_iter()
+                    .map(|blob_id| self.storage.read_hashed_blob(blob_id)),
```

Comment:
If we are sending a certificate other than `ConfirmedBlock`, the blob usually won't be in storage yet. But it should be in our local node's chain manager.

Reply:
In `ChainClient::propose_block` we first do `handle_block_proposal` in our own local node, so the local node's chain manager should now have the blob.

On the other hand, there are also cases where we need to re-submit a validated block based on another owner's proposal. In that case, _only_ our chain manager will have the blob, but not the client's pending blobs.

In general: The updater is trying to update the validators about something that we already updated our local node about, so it should always have the blobs. Thus I don't think we need to access the client's pending blobs here.